### DISCUSSION TITLE
Use-cannon and Socket.io ?

### DISCUSSION THREAD

**Discussion by Franckapik** [pmndrs/use-cannon#74] [General] [answered]
I have a game with one plane and one sphere per player. 
If a player press the arrows key, a applyLocalForce is applied on its sphere.
I use socket.io to share from server the keypress (true/false) between clients.
Now, when i have two opened windows, i see some differences of position between both views. 
I know maybe i have to read about fps and socket. But my question is : 

-Do i have to share the position from subscribed position or the key event to apply forces on each client ? Is it possible to have interpolation (with lerp) with physics impulses?

Thank you a lot for your help !

F.

**Comment by codynova** [comment]
I'm honestly not familiar with the best approach to synchronizing the physics engine in a multiplayer environment. I'd need to do some research before I could recommend an appropriate approach.

**Comment by stockhuman** [comment]
That's a very big question - from what I can tell, cannon is non-deterministic in real time. See [this exchange](https://stackoverflow.com/questions/39218524/aframe-physics-impulse-is-not-at-center) with Don McCurdy and schteppe's [forum post](https://www.html5gamedevs.com/topic/22996-what-are-physical-equations-used-by-cannonjs-and-how-to-map-them-onto-a-babylon-code/?do=findComment&comment=131437).

This [article](https://medium.com/madebywild/building-a-realtime-multiplayer-game-with-deepstream-c80d6e1bc50a) (and project) might help you.

**Comment by codynova** [comment, marked as answer]
Rather than send events (key presses) across the network and calculating the results in each client, you should run `cannon-es` on the server. Send the events to the server, let it calculate the results, then distribute those results back to the client.

Be aware that there is currently (`cannon-es@0.10.0`) a small bug with running `cannon-es` in Node, which I'm going to try to look at today: https://github.com/react-spring/cannon-es/issues/23#issuecomment-619619159

EDIT: I should have mentioned that the current implementation of `use-cannon` does not support using `cannon-es` in this way.

**Comment by Franckapik** [comment]
Thank you @codynova for your work !
So, it could be a great idea to use cannon-es from the server (using Express/Node.js) to make calculation server-side.
Indeed, i have the performance={} issue :/ I am seeing that your are working on it this time... So i will take a look theses days to see if you purpose resolution for this issue...

**Comment by Franckapik** [comment]
I have used the little correction to avoid the "performance issue".
I am now just a little confused of using cannon on server.
Do you have an idea of what i should emit from server to client to update position of a mesh in react three fiber ? 
Here is the server socket page when i try to input cannon-es  : 
https://github.com/Franckapik/NandY/blob/master/routes/socket.js
When i read the docs i see that  Impulses will be added to Body.velocity and Body.angularVelocity. Do i have to share the body object ? So i should keep use-cannon on client ?

Sorry for this big question but your response will be a huge help for me !

F.

**Comment by codynova** [comment]
Sorry, I should have clarified: the current implementation of `use-cannon` will not allow you to run `cannon-es` on the server. I've never even tried running `cannon-es` from the server, but I believe the solution I described is theoretically possible with the current `cannon-es` implementation. I will consider how `use-cannon` could be changed to allow for this pattern, but don't expect it any time soon (unfortunately).

**Reply by Franckapik** [reply]
Ok! So what you suggested was to try to use cannon-es on server side only and affect position of object by emitting position of the player to the client ? 
I will try to see if there is some example of cannon.js on server side to see if it's possible to do this with my configuration.
But if it's too hard, unfortunately, i will use use-cannon only and forget about using multiplayer. :(

**Reply by codynova** [reply]
>So what you suggested was to try to use cannon-es on server side only and affect position of object by emitting position of the player to the client ?

Yep that's more or less what I'd try - have the clients establish a socket connection to the server running the World simulation, and receive updates through there.

**Comment by Franckapik** [comment]
I tried to make a sphere falling on the ground and with impulse from a key event on server side.
i use world.step in a gameloop and send to the client the position of the sphere each frame. With 60 fps, i have big latency and it seem's that it's already too much... :/
I am wrong somewhere in this code ? I know that socket in this situation is particular but if someone could help me on this ... :)

https://github.com/Franckapik/NandY/blob/master/routes/socket.js
